pkg/csi-handler/pv_checker.go

```diff
@@ -0,0 +1,223 @@
+	defer cancel()
+
+	result, err := checker.csiPVHandler.ControllerVolumesChecking(ctx)
+	if err != nil {
+		return err
+	}
+
+	pvs, err := checker.pvLister.List(labels.Everything())
+	if err != nil {
+		return err
+	}
+
+	for _, pv := range pvs {
+		if pv.Spec.CSI == nil || pv.Spec.CSI.Driver != checker.checkerName {
+			klog.Infof("csi source is nil or the volume is not managed by this checker/monitor")
+			continue
+		}
+
+		if pv.Status.Phase != v1.VolumeBound {
+			klog.Infof("PV: %s status is not bound", pv.Name)
+			continue
+		}
+
+		volumeHandle, err := checker.GetVolumeHandle(pv)
+		if err != nil {
+			klog.Errorf("Get volume handler error: %+v", err)
+			continue
+		}
+
+		if result[volumeHandle] != nil && result[volumeHandle].GetAbnormal() {
+			// Since pv status is bound, we believe PV controller, do not check pv.Spec.ClaimRef here.
+			pvc, err := checker.pvcLister.PersistentVolumeClaims(pv.Spec.ClaimRef.Namespace).Get(pv.Spec.ClaimRef.Name)
+			if err != nil {
+				klog.Errorf("Get PVC error: %+v", err)
+				continue
+			}
+			checker.eventRecorder.Event(pvc, v1.EventTypeWarning, "PV is not healthy(detected by monitor controller)", result[volumeHandle].GetMessage())
+		}
+	}
+
+	return nil
+}
+
+// GetVolumeHandle return the volume handle of the pv
+func (checker *PVHealthConditionChecker) GetVolumeHandle(pv *v1.PersistentVolume) (string, error) {
+	if pv.Spec.CSI == nil {
+		return "", fmt.Errorf("csi source is nil")
+	}
+
+	return pv.Spec.CSI.VolumeHandle, nil
+}
+
+// CheckControllerVolumeStatus checks volume status in controller side
+func (checker *PVHealthConditionChecker) CheckControllerVolumeStatus(pv *v1.PersistentVolume) error {
+	if pv.Spec.CSI == nil || pv.Spec.CSI.Driver != checker.checkerName {
+		return fmt.Errorf("csi source is nil or the volume is not managed by this checker/monitor")
+	}
+
+	if pv.Status.Phase != v1.VolumeBound {
+		return fmt.Errorf("PV: %s status is not bound", pv.Name)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), checker.timeout)
+	defer cancel()
+
+	volumeHandle, err := checker.GetVolumeHandle(pv)
+	if err != nil {
+		klog.Errorf("Get volume handler error: %+v", err)
+		return err
+	}
+
+	if len(volumeHandle) == 0 {
+		return fmt.Errorf("volume handle in csi source is empty")
+	}
+
+	volumeCondition, err := checker.csiPVHandler.ControllerVolumeChecking(ctx, volumeHandle)
+	if err != nil {
+		return err
+	}
+
+	// if error is not nil, the volumeCondition result can not be nil neither
+
+	// At the first stage, we just send PVC events
+	if volumeCondition.GetAbnormal() {
+		// Since pv status is bound, we believe PV controller, do not check pv.Spec.ClaimRef here.
+		pvc, err := checker.pvcLister.PersistentVolumeClaims(pv.Spec.ClaimRef.Namespace).Get(pv.Spec.ClaimRef.Name)
+		if err != nil {
+			return err
+		}
+		checker.eventRecorder.Event(pvc, v1.EventTypeWarning, "PV is not healthy(detected by monitor controller)", volumeCondition.GetMessage())
```

Comment:
s/PV is not healthy(detected by monitor controller)/Volume condition is abnormal

Reply:
done